### PR TITLE
Split `SolverDelegate` back out from `InferCtxtLike`

### PR DESCRIPTION
**Pull request by compiler-errors**
This is because in order to uplift things like the `Generalizer` and other `TypeRelation`s, we want to be able to interface with `InferCtxtLike` (and `InferCtxt` as its implementation), rather that `SolverDelegate`, which only really exists as a hack to be able to define some downstream methods in `rustc_type_ir`.

r? lcnr

### CODE REVIEW

File: compiler/rustc_next_trait_solver/src/delegate.rs [hunk 1]
```diff
@@ -1,18 +1,20 @@
...
 
 use rustc_type_ir::fold::TypeFoldable;
-use rustc_type_ir::relate::Relate;
 use rustc_type_ir::solve::{Certainty, Goal, NoSolution, SolverMode};
-use rustc_type_ir::{self as ty, Interner};
+use rustc_type_ir::{self as ty, InferCtxtLike, Interner};
 
-pub trait SolverDelegate: Sized {
+pub trait SolverDelegate:
+    Deref<Target: InferCtxtLike<Interner = <Self as SolverDelegate>::Interner>> + Sized
```
Comment (compiler-errors) [inline compiler/rustc_next_trait_solver/src/delegate.rs:9]: This is the bound that lets us use a `SolverDelegate` like an `InferCtxtLike`.

File: compiler/rustc_infer/src/infer/context.rs [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+    fn create_next_universe(&self) -> ty::UniverseIndex {
+        self.create_next_universe()
+    }
+
+    fn universe_of_ty(&self, vid: ty::TyVid) -> Option<ty::UniverseIndex> {
+        // FIXME(BoxyUwU): this is kind of jank and means that printing unresolved
+        // ty infers will give you the universe of the var it resolved to not the universe
+        // it actually had. It also means that if you have a `?1` and infer it to `u8` then
+        // try to print out `?1` it will just print `?0`.
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/context.rs:39]: that fixme is outdated now, isn't it?

File: compiler/rustc_infer/src/infer/context.rs [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+    fn universe_of_lt(&self, lt: ty::RegionVid) -> Option<ty::UniverseIndex> {
+        match self.inner.borrow_mut().unwrap_region_constraints().probe_value(lt) {
+            Err(universe) => Some(universe),
+            Ok(_) => None,
+        }
+    }
+
+    fn universe_of_ct(&self, ct: ty::ConstVid) -> Option<ty::UniverseIndex> {
+        // Same issue as with `universe_of_ty`
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/context.rs:54]: and here

**Review by lcnr — COMMENTED** [review]
r=me after nit

**Comment by compiler-errors** [comment]
They very much are not. Why didn't rustbot comment that submodules were changed :(

**Comment by compiler-errors** [comment]
OK, reverted the submodule changes.

@bors r=lcnr

**Pull request merged**